Mimir/Controllers/AvatarController.cs

```diff
@@ -1,23 +1,136 @@
+        }
+    }
+
+    // This is a snippet from Mimir/Util/StateJsonConverter.cs
+    private class StateJsonConverter : JsonConverter
+    {
+        public override bool CanConvert(Type objectType)
+        {
+            return typeof(IState).IsAssignableFrom(objectType);
+        }
+
+        public override object ReadJson(
+            JsonReader reader,
+            Type objectType,
+            object? existingValue,
+            JsonSerializer serializer)
+        {
+            throw new NotImplementedException();
+        }
+
+        public override void WriteJson(JsonWriter writer, object? value, JsonSerializer serializer)
+        {
+            if (value is null)
+            {
+                return;
+            }
+
+            var jo = JObject.FromObject(value, JsonSerializer.CreateDefault(new JsonSerializerSettings
+            {
+                ReferenceLoopHandling = ReferenceLoopHandling.Ignore,
+                Converters = new List<JsonConverter> { new BigIntegerToStringConverter() },
+                Formatting = Formatting.Indented,
+                NullValueHandling = NullValueHandling.Ignore
+            }));
+
+            var iValue = value switch
+            {
+                AvatarState avatarState => avatarState.SerializeList(),
+                IState state => state.Serialize(),
+                _ => null
+            };
+
+            if (iValue != null)
+            {
+                var rawValue = ByteUtil.Hex(new Codec().Encode(iValue));
+                jo.Add("Raw", rawValue);
+            }
+
+            jo.WriteTo(writer);
+        }
+    }
+
+    // This is a snippet from Mimir.Worker/Models/State/BaseData.cs
+    private static JsonSerializerSettings JsonSerializerSettings => new()
+    {
+        Converters = { new StateJsonConverter() },
+        Formatting = Formatting.Indented,
+        NullValueHandling = NullValueHandling.Ignore
+    };
+
+    #endregion snippets
+
     [HttpGet("{avatarAddress}/inventory")]
-    public Inventory? GetInventory(string network, string avatarAddress)
+    public async Task<Inventory?> GetInventory(
+        string network,
+        string avatarAddress,
+        IStateService stateService)
     {
         var inventory = avatarRepository.GetInventory(network, avatarAddress);
-        if (inventory is null)
+        if (inventory is not null)
+        {
+            return inventory;
+        }
+
+        var stateGetter = new StateGetter(stateService);
+        var inventoryState = await stateGetter.GetInventoryStateAsync(new Address(avatarAddress));
+        if (inventoryState is null)
         {
             Response.StatusCode = StatusCodes.Status404NotFound;
             return null;
         }
 
+        try
+        {
+            var jsonString = JsonConvert.SerializeObject(inventoryState, JsonSerializerSettings);
+            var bsonDocument = BsonDocument.Parse(jsonString);
+            inventory = new Inventory(bsonDocument);
```

Comment:
It seems like unnecessary code which can cause performance issues.  You shouldn't have any problems with the behavior, but if you do, this is the place to fix it. I guess that it can be solved by overloading the class constructor.

Reply:
I think you probably have a misunderstanding about the `Inventory` type in `Mimir` and the `Inventory` type in `Lib9c`.
The `Inventory` type on line 126 is a `Mimir` project-specific type.
https://github.com/planetarium/mimir/blob/main/Mimir/Models/Avatar/Inventory.cs

Reply:
Thanks for the kind explanation, but I wasn't confusing the two. What I meant to say was Lib9c.Inventory -> JSON -> BSON -> Mimir.Inventory, so why not just interpret it as Lib9c.Inventory -> Mimir.Inventory?

Reply:
Oh right, we don't have the ability to convert Lib9c.Inventory to Mimir.Inventory yet, let's add it here.